visor/visor.go

```diff
@@ -0,0 +1,199 @@
+	r.log.Debug("setting current folder",
+		logging.String("sourceFolder", sourceFolder),
+		logging.String("currentFolder", currentFolder),
+	)
+
+	runConfPath := path.Join(sourceFolder, config.RunConfigFileName)
+	runConfExists, err := utils.PathExists(runConfPath)
+	if err != nil {
+		return err
+	}
+
+	if !runConfExists {
+		return fmt.Errorf("missing run config in %q folder", runConfPath)
+	}
+
+	if err := os.RemoveAll(currentFolder); err != nil {
+		return fmt.Errorf("failed to remove current folder: %w", err)
+	}
+
+	if err := os.Symlink(sourceFolder, currentFolder); err != nil {
+		return fmt.Errorf("failed to set current folder as current: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Visor) Run(ctx context.Context) error {
+	numOfRestarts := 0
+
+	upgradeTicker := time.NewTicker(upgradeApiCallTickerDuration)
+	defer upgradeTicker.Stop()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		runConf, err := config.ParseRunConfig(r.conf.CurrentRunConfigPath())
+		if err != nil {
+			return fmt.Errorf("failed to parse run config: %w", err)
+		}
+
+		client := r.clientFactory.GetClient(runConf.RCP.SocketPath, runConf.RCP.HttpPath)
+
+		numOfUpgradeStatusErrs := 0
+		maxNumRestarts := r.conf.MaxNumberOfRestarts()
+		restartsDelay := time.Second * time.Duration(r.conf.RestartsDelaySeconds())
+
+		r.log.Info("Starting binaries")
+		binRunner := NewBinariesRunner(r.log, r.conf.CurrentFolder(), time.Second*time.Duration(r.conf.StopSignalTimeoutSeconds()))
+		binErrs := binRunner.Run(ctx, runConf.Binaries)
+
+		upgradeTicker.Reset(upgradeApiCallTickerDuration)
+
+	CheckLoop:
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case err := <-binErrs:
+				r.log.Error("Binaries executions has failed", logging.Error(err))
+
+				if numOfRestarts >= maxNumRestarts {
+					return fmt.Errorf("maximum number of possible restarts has been reached: %w", err)
+				}
+
+				numOfRestarts++
+				r.log.Info("Binaries restart is scheduled", logging.Duration("restartDelay", restartsDelay))
+				time.Sleep(restartsDelay)
+				r.log.Info("Restarting binaries", logging.Int("remainingRestarts", maxNumRestarts-numOfRestarts))
+
+				break CheckLoop
+			case <-upgradeTicker.C:
+				upStatus, err := client.UpgradeStatus(ctx)
+				if err != nil {
+					if numOfUpgradeStatusErrs > maxUpgradeStatusErrs {
+						return fmt.Errorf("failed to upgrade status for maximum amount of %d times: %w", maxUpgradeStatusErrs, err)
+					}
+
+					r.log.Debug("failed to get upgrade status from API", logging.Error(err))
+					numOfUpgradeStatusErrs++
+
+					break
+				}
+
+				if !upStatus.ReadyToUpgrade {
+					break
+				}
+
+				r.log.Info("Preparing upgrade")
+
+				if err := binRunner.Stop(); err != nil {
+					// Force to kill if fails grateful stop fails
+					if err := binRunner.Kill(); err != nil {
+						return fmt.Errorf("failed to force kill the running processes: %w", err)
+					}
+				}
+
+				r.log.Info("Starting upgrade")
```

Comment:
but what starts the new binaries?

Reply:
We set the upgrade folder (the one with the new binaries) as current. So with the next loop iteration we are running a new (updated) binaries.

Reply:
ah got it, thanks